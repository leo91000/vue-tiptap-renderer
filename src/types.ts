export { JSONContent } from '@tiptap/core'
