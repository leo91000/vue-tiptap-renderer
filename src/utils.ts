import type { JSONContent } from './types'

export function getContentFirstImage(content: JSONContent): { src: string, alt: string, title: string } | null {
  if (content.type === 'image') {
    return {
      src: content.attrs.src,
      alt: content.attrs.alt,
      title: content.attrs.title,
    }
  }

  if (content.content) {
    for (const child of content.content) {
      const firstImage = getContentFirstImage(child)
      if (firstImage)
        return firstImage
    }
  }

  return null
}

export function getContentWordCount(content: JSONContent): number {
  let wordCount = 0

  if (!content)
    return wordCount

  if (content.content) {
    for (const child of content.content)
      wordCount += getContentWordCount(child)
  }

  if (content.text)
    wordCount += content.text.split(' ').length

  return wordCount
}

export function extractContentText(content: JSONContent): string {
  let text = ''

  if (!content)
    return text

  if (content.content) {
    for (const child of content.content)
      text += extractContentText(child)
  }

  if (content.text)
    text += `${content.text}\n`

  return text
}

export const emptyJsonContent: JSONContent = {
  type: 'doc',
  content: [{
    type: 'paragraph',
    content: [{
      type: 'text',
      text: '',
    }],
  }],
}
