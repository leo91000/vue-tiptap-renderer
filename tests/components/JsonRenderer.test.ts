import { describe, expect, it } from 'vitest'
import { mount } from '@vue/test-utils'
import { defineComponent } from 'vue'
import type { JSONContent } from '../../src'
import { JsonRenderer } from '../../src'

const WrappedJsonRenderer = defineComponent({
  components: {
    JsonRenderer,
  },
  props: {
    content: {
      type: Object as () => JSONContent,
      required: true,
    },
  },
  template: '<JsonRenderer v-bind="$props" />',
})

describe('jsonRenderer', () => {
  it('should render text', () => {
    const wrapper = mount(WrappedJsonRenderer, {
      props: {
        content: {
          type: 'doc',
          content: [{ type: 'text', text: 'ARTICLE_RENDERED_SUCCESSFULLY' }],
        },
      },
    })

    expect(wrapper.html()).toContain('ARTICLE_RENDERED_SUCCESSFULLY')
  })

  it('should render complex articles', () => {
    const wrapper = mount(WrappedJsonRenderer, {
      props: {
        content: { type: 'doc', content: [{ type: 'paragraph', attrs: { textAlign: 'left' } }, { type: 'image', attrs: { alt: 'Sans titre.jpg', src: 'https://aqhptogkwxnadanqnhgi.supabase.co/storage/v1/object/public/storage/20220501165812941_Sans titre.jpg', title: 'Sans titre.jpg' } }, { type: 'paragraph', attrs: { textAlign: 'left' }, content: [{ text: ' ', type: 'text', marks: [{ type: 'bold' }, { type: 'textStyle', attrs: { fontFamily: '' } }] }] }, { type: 'paragraph', attrs: { textAlign: 'justify' }, content: [{ text: 'Être aimé, est à l’enfance une question de survie. Les yeux et oreilles rivés vers l’extérieur, vous entendez les injonctions et expressions d’autrui. D’abord vous écoutez, et puis vous vous adaptez.', type: 'text', marks: [{ type: 'textStyle', attrs: { fontFamily: 'Calibri, sans-serif' } }] }] }, { type: 'paragraph', attrs: { textAlign: 'justify' }, content: [{ type: 'hardBreak', marks: [{ type: 'textStyle', attrs: { fontFamily: 'Calibri, sans-serif' } }] }, { text: 'Ce besoin indispensable d’être aimé peut pourtant devenir la source de problèmes aussi communs qu’importants. Car même s’il est temporaire, le conditionnement reste. Dans ce cas, votre attention est focalisée vers les autres. Vous êtes donc alerte aux choses du monde (les pensées probables d’autrui, ce qu’il est préférable de faire ou non…) : adepte du compromis et de la comparaison. Et souvent, c’est inconscient. Il est possible que vous agissiez en considérant l’extérieur en premier. Cette course à la vie, à la survie parfois, peut vous faire oublier l’essentiel : ', type: 'text', marks: [{ type: 'textStyle', attrs: { fontFamily: 'Calibri, sans-serif' } }] }, { text: 'et moi ? ', type: 'text', marks: [{ type: 'bold' }, { type: 'italic' }, { type: 'textStyle', attrs: { fontFamily: 'Calibri, sans-serif' } }] }, { text: 'Car la responsabilité des actes envers Soi existe et devrait compter autant que celle des actes envers l’Extérieur ', type: 'text', marks: [{ type: 'textStyle', attrs: { fontFamily: 'Calibri, sans-serif' } }] }] }, { type: 'paragraph', attrs: { textAlign: 'justify' }, content: [{ type: 'hardBreak', marks: [{ type: 'textStyle', attrs: { fontFamily: 'Calibri, sans-serif' } }] }, { text: 'RENTRER EN RELATION AVEC SOI MÊME ', type: 'text', marks: [{ type: 'bold' }, { type: 'textStyle', attrs: { fontFamily: 'Calibri, sans-serif' } }] }, { type: 'hardBreak', marks: [{ type: 'bold' }, { type: 'textStyle', attrs: { fontFamily: 'Calibri, sans-serif' } }] }, { text: 'C’est incontournable, pour commencer à s’aimer d’avantage il faut ', type: 'text', marks: [{ type: 'textStyle', attrs: { fontFamily: 'Calibri, sans-serif' } }] }, { text: 'le décider', type: 'text', marks: [{ type: 'bold' }, { type: 'textStyle', attrs: { fontFamily: 'Calibri, sans-serif' } }] }, { text: '. C’est-à-dire, décider que vous serez la première personne sur qui vous pouvez compter. Créez un espace intérieur pour vous soucier de vous. Encouragez-vous, aidez-vous.', type: 'text', marks: [{ type: 'textStyle', attrs: { fontFamily: 'Calibri, sans-serif' } }] }, { type: 'hardBreak', marks: [{ type: 'textStyle', attrs: { fontFamily: 'Calibri, sans-serif' } }] }, { text: 'Évaluer l’état de cette relation à vous-même, peut être un vrai plus. Comment est-ce je me parle et me vois ?', type: 'text', marks: [{ type: 'textStyle', attrs: { fontFamily: 'Calibri, sans-serif' } }] }] }, { type: 'paragraph', attrs: { textAlign: 'justify' }, content: [{ text: ' ', type: 'text', marks: [{ type: 'textStyle', attrs: { fontFamily: 'Calibri, sans-serif' } }] }, { type: 'hardBreak', marks: [{ type: 'textStyle', attrs: { fontFamily: 'Calibri, sans-serif' } }] }, { text: 'CONSCIENT DE SES MOTIVATIONS', type: 'text', marks: [{ type: 'bold' }, { type: 'textStyle', attrs: { fontFamily: 'Calibri, sans-serif' } }] }, { type: 'hardBreak', marks: [{ type: 'textStyle', attrs: { fontFamily: 'Calibri, sans-serif' } }] }, { text: 'Nos actions ont parfois un rapport avec soi. Par exemple ; j’apporte un café à une collègue pour lui faire plaisir. Il est vrai que je lui apporte ce sentiment. Mais je le fais aussi parce que ça me procure de la joie. Prendre conscience que vous êtes deux à bénéficier de cet acte, vous feras toucher du doigt ce que peut être un acte d’amour envers soi. ', type: 'text', marks: [{ type: 'textStyle', attrs: { fontFamily: 'Calibri, sans-serif' } }] }] }, { type: 'paragraph', attrs: { textAlign: 'justify' }, content: [{ type: 'hardBreak', marks: [{ type: 'textStyle', attrs: { fontFamily: 'Calibri, sans-serif' } }] }, { text: 'DES PETITS MOMENTS', type: 'text', marks: [{ type: 'bold' }, { type: 'textStyle', attrs: { fontFamily: 'Calibri, sans-serif' } }] }, { type: 'hardBreak', marks: [{ type: 'textStyle', attrs: { fontFamily: 'Calibri, sans-serif' } }] }, { text: 'Un petit tête-à-tête ? Oui, mais avec vous-même de temps en temps : c’est mieux ! Prenez ces petits moments, écrivez dans un journal ou juste pensez, accordez-vous un moment pour vous exprimer entre vous et vous. Et faite des pauses, prenez tout simplement des moments pour vous faire plaisir.', type: 'text', marks: [{ type: 'textStyle', attrs: { fontFamily: 'Calibri, sans-serif' } }] }] }, { type: 'paragraph', attrs: { textAlign: 'justify' }, content: [{ type: 'hardBreak', marks: [{ type: 'textStyle', attrs: { fontFamily: 'Calibri, sans-serif' } }] }, { text: 'ACCUEILLEZ VOS PROPRES ÉMOTIONS', type: 'text', marks: [{ type: 'bold' }, { type: 'textStyle', attrs: { fontFamily: 'Calibri, sans-serif' } }] }, { text: ' ', type: 'text', marks: [{ type: 'textStyle', attrs: { fontFamily: 'Calibri, sans-serif' } }] }, { type: 'hardBreak', marks: [{ type: 'textStyle', attrs: { fontFamily: 'Calibri, sans-serif' } }] }, { text: 'Comment ça va moi ? Qu’est que je ressens ? Les défis et les événements du quotidien créent des réactions intérieures. Parfois pour des sujets qui paraissent anodins. Vous n’êtes pas obligé d’attendre l’aide de quelqu’un pour vous exprimer. Vous avez le droit de cogner dans un coussin ou de pleurer quand cela est bon ! Relâcher l’émotion peut amener à la compréhension de quelque chose qui se joue en vous. Cette habitude peut aussi améliorer vos relations, en vous évitant d’avantage d’exploser ou de projeter sur les autres. ', type: 'text', marks: [{ type: 'textStyle', attrs: { fontFamily: 'Calibri, sans-serif' } }] }] }, { type: 'paragraph', attrs: { textAlign: 'justify' }, content: [{ type: 'hardBreak', marks: [{ type: 'textStyle', attrs: { fontFamily: 'Calibri, sans-serif' } }] }, { text: 'DÉCIDEZ POUR VOUS', type: 'text', marks: [{ type: 'bold' }, { type: 'textStyle', attrs: { fontFamily: 'Calibri, sans-serif' } }] }, { type: 'hardBreak', marks: [{ type: 'bold' }, { type: 'textStyle', attrs: { fontFamily: 'Calibri, sans-serif' } }] }, { text: 'Ne laissez pas les autres décider pour vous. À la fin, c’est vous qui avez le dernier mot sur ce à quoi doit ressembler votre vie ou votre quotidien. Soyez conscient de votre part de responsabilité dans la vie que vous vivez. Attention responsabilité ne veut pas dire culpabilité. ', type: 'text', marks: [{ type: 'textStyle', attrs: { fontFamily: 'Calibri, sans-serif' } }] }] }, { type: 'paragraph', attrs: { textAlign: 'justify' }, content: [{ type: 'hardBreak', marks: [{ type: 'textStyle', attrs: { fontFamily: 'Calibri, sans-serif' } }] }, { text: 'PARDONNEZ-VOUS LES « FAUSSES ROUTES »', type: 'text', marks: [{ type: 'bold' }, { type: 'textStyle', attrs: { fontFamily: 'Calibri, sans-serif' } }] }, { type: 'hardBreak', marks: [{ type: 'textStyle', attrs: { fontFamily: 'Calibri, sans-serif' } }] }, { text: '« Avec la culpabilité, le malheur est la chose la plus démocratique du monde. On y a tous droit à un moment ou un autre », Eric Neuhoff.', type: 'text', marks: [{ type: 'textStyle', attrs: { fontFamily: 'Calibri, sans-serif' } }] }, { type: 'hardBreak', marks: [{ type: 'textStyle', attrs: { fontFamily: 'Calibri, sans-serif' } }] }, { text: 'Nourrir un sentiment négatif à votre égard favorisera comme cela le ferait avec une personne extérieure, un éloignement. Il peut devenir la cause de nombreux maux du quotidien. Si vos sentiments ètaient une valise, la culpabilité peut y être l’objet le plus lourd. Cette valise si lourde, vous permettra elle de voyager à votre guise ?', type: 'text', marks: [{ type: 'textStyle', attrs: { fontFamily: 'Calibri, sans-serif' } }] }] }, { type: 'paragraph', attrs: { textAlign: 'justify' }, content: [{ type: 'hardBreak', marks: [{ type: 'textStyle', attrs: { fontFamily: 'Calibri, sans-serif' } }] }, { text: 'STOP A LA COMPARAISON', type: 'text', marks: [{ type: 'bold' }, { type: 'textStyle', attrs: { fontFamily: 'Calibri, sans-serif' } }] }, { type: 'hardBreak', marks: [{ type: 'textStyle', attrs: { fontFamily: 'Calibri, sans-serif' } }] }, { text: 'Ne vous focalisez pas sur ce que l’autre a et que vous n’avez pas. Cela peut valider et nourrir l’idée que vous avez de ne pas être pas à la hauteur ou que vous êtes moins bien, etc. Partez à la découverte de ce qui est beau chez vous, de vos qualités, de vos talents, de vos chances…', type: 'text', marks: [{ type: 'textStyle', attrs: { fontFamily: 'Calibri, sans-serif' } }] }, { type: 'hardBreak', marks: [{ type: 'textStyle', attrs: { fontFamily: 'Calibri, sans-serif' } }] }] }] },
      },
    })

    expect(
      wrapper.html(),
    ).toMatchSnapshot()
  })
})
