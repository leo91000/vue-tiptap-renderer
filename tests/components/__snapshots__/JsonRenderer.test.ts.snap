// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`jsonRenderer > should render complex articles 1`] = `
"<p style="text-align: left;"></p>
<img src="https://aqhptogkwxnadanqnhgi.supabase.co/storage/v1/object/public/storage/20220501165812941_Sans titre.jpg" alt="Sans titre.jpg" title="Sans titre.jpg">
<p style="text-align: left;"><strong>&nbsp;</strong></p>
<p style="text-align: justify;"><span style="font-family: Calibri, sans-serif;">Être aimé, est à l’enfance une question de survie.&nbsp;Les yeux et oreilles rivés vers l’extérieur, vous entendez les injonctions et expressions d’autrui.&nbsp;D’abord vous&nbsp;écoutez, et puis vous vous adaptez.</span></p>
<p style="text-align: justify;"><br><span style="font-family: Calibri, sans-serif;">Ce besoin indispensable d’être aimé peut pourtant devenir la source de problèmes aussi communs qu’importants.&nbsp;Car même s’il est temporaire, le conditionnement reste.&nbsp;Dans ce cas, votre attention est focalisée vers les autres.&nbsp;Vous êtes donc alerte aux choses du monde&nbsp;(les pensées probables d’autrui, ce qu’il est préférable de faire ou non…)&nbsp;:&nbsp;adepte du compromis et de la comparaison.&nbsp;Et souvent, c’est inconscient.&nbsp;Il est possible que vous agissiez en considérant l’extérieur en premier.&nbsp;Cette course à la vie, à la&nbsp;survie parfois, peut vous faire oublier l’essentiel :&nbsp;</span><i style="font-family: Calibri, sans-serif;"><strong>et moi ? </strong></i><span style="font-family: Calibri, sans-serif;">Car la responsabilité des actes envers&nbsp;Soi&nbsp;existe et devrait compter autant que celle des actes envers l’Extérieur&nbsp;</span></p>
<p style="text-align: justify;"><br><strong style="font-family: Calibri, sans-serif;">RENTRER EN RELATION AVEC SOI MÊME&nbsp;</strong><br><span style="font-family: Calibri, sans-serif;">C’est incontournable, pour commencer à s’aimer d’avantage il&nbsp;faut </span><strong style="font-family: Calibri, sans-serif;">le décider</strong><span style="font-family: Calibri, sans-serif;">.&nbsp;C’est-à-dire, décider que vous serez la première personne sur qui vous pouvez compter.&nbsp;Créez un espace intérieur pour vous soucier de vous.&nbsp;Encouragez-vous, aidez-vous.</span><br><span style="font-family: Calibri, sans-serif;">Évaluer l’état de cette relation à vous-même, peut être un vrai plus.&nbsp;Comment est-ce je me parle et me vois ?</span></p>
<p style="text-align: justify;"><span style="font-family: Calibri, sans-serif;">&nbsp;</span><br><strong style="font-family: Calibri, sans-serif;">CONSCIENT DE SES MOTIVATIONS</strong><br><span style="font-family: Calibri, sans-serif;">Nos actions ont parfois un rapport avec soi.&nbsp;Par exemple ;&nbsp;j’apporte un café à une collègue pour lui faire plaisir.&nbsp;Il est vrai que je lui apporte ce sentiment.&nbsp;Mais je le fais aussi parce que ça me procure de la joie.&nbsp;Prendre conscience que vous êtes deux à bénéficier de cet acte, vous feras toucher du doigt ce que peut être un acte d’amour envers soi.&nbsp;</span></p>
<p style="text-align: justify;"><br><strong style="font-family: Calibri, sans-serif;">DES PETITS MOMENTS</strong><br><span style="font-family: Calibri, sans-serif;">Un petit&nbsp;tête-à-tête ?&nbsp;Oui, mais avec vous-même de temps en temps :&nbsp;c’est mieux !&nbsp;Prenez ces petits moments, écrivez dans un journal ou juste pensez, accordez-vous un moment pour vous exprimer entre vous et vous.&nbsp;Et faite des pauses, prenez tout simplement des moments pour vous faire plaisir.</span></p>
<p style="text-align: justify;"><br><strong style="font-family: Calibri, sans-serif;">ACCUEILLEZ VOS PROPRES ÉMOTIONS</strong><span style="font-family: Calibri, sans-serif;">&nbsp;</span><br><span style="font-family: Calibri, sans-serif;">Comment ça&nbsp;va moi ?&nbsp;Qu’est que je ressens ?&nbsp;Les défis et les événements du quotidien créent des réactions intérieures.&nbsp;Parfois pour des sujets qui paraissent anodins.&nbsp;Vous n’êtes pas obligé d’attendre l’aide de quelqu’un pour vous exprimer.&nbsp;Vous avez le droit de cogner dans un coussin ou de pleurer quand cela est bon !&nbsp;Relâcher l’émotion peut amener à la compréhension de quelque chose qui se joue en vous.&nbsp;Cette&nbsp;habitude peut aussi améliorer vos relations, en vous évitant d’avantage d’exploser ou de projeter sur les autres.&nbsp;</span></p>
<p style="text-align: justify;"><br><strong style="font-family: Calibri, sans-serif;">DÉCIDEZ POUR VOUS</strong><br><span style="font-family: Calibri, sans-serif;">Ne laissez pas les autres décider pour vous.&nbsp;À la fin, c’est vous qui avez le dernier mot sur ce à quoi doit ressembler votre vie ou votre quotidien.&nbsp;Soyez conscient de votre part de responsabilité dans la vie que vous vivez.&nbsp;Attention responsabilité ne veut pas dire culpabilité.&nbsp;</span></p>
<p style="text-align: justify;"><br><strong style="font-family: Calibri, sans-serif;">PARDONNEZ-VOUS LES « FAUSSES ROUTES »</strong><br><span style="font-family: Calibri, sans-serif;">« Avec la culpabilité, le malheur est la chose la plus démocratique du monde.&nbsp;On y a tous droit à un moment ou un autre », Eric Neuhoff.</span><br><span style="font-family: Calibri, sans-serif;">Nourrir un sentiment négatif à votre égard favorisera comme cela le ferait avec une personne extérieure, un éloignement.&nbsp;Il peut devenir la cause de nombreux maux du quotidien. Si vos sentiments ètaient une valise,&nbsp;la culpabilité peut y être l’objet le plus lourd.&nbsp;Cette valise si lourde, vous permettra elle de voyager à votre guise ?</span></p>
<p style="text-align: justify;"><br><strong style="font-family: Calibri, sans-serif;">STOP A LA COMPARAISON</strong><br><span style="font-family: Calibri, sans-serif;">Ne vous focalisez pas sur ce que l’autre a&nbsp;et&nbsp;que vous n’avez pas.&nbsp;Cela peut valider et nourrir l’idée que vous avez de ne pas être pas à la hauteur ou que vous êtes moins bien, etc.&nbsp;Partez à la découverte de ce qui est beau chez vous, de vos qualités, de vos talents, de vos chances…</span><br></p>"
`;
